gnuplot/src/color.rs

```diff
@@ -0,0 +1,413 @@
+	Index(ColorIndex),
+	/// A color type that sets the color per element using a index `n` which represents the `n`th
+	/// color in the current gnuplot color scheme. In gnuplot this is the last element in the plot command,
+	/// in Rust gnuplot, the color type takes a vector of u8, where each index is treated the same as the
+	/// fixed `IndexColor`.
+	/// This is useful for setting bars/boxes etc to be
+	/// the same color from multiple plot commands. The `variable_color` example has examples of this usage.
+	VariableIndex(Vec<ColorIndex>),
+	/// Set the color of the plot to the current background color.
+	Background,
+	/// Fixed black color
+	Black,
+}
+
+impl<T: Display> ColorType<T>
+{
+	/// Returns the gnuplot string that will produce the requested color
+	pub fn command(&self) -> String
+	{
+		let str = match self
+		{
+			RGBString(s) => &format!(r#"rgb "{}""#, s),
+			RGBInteger(r, g, b) => &format!(r#"rgb {}"#, from_argb(0, *r, *g, *b)),
+			ARGBInteger(a, r, g, b) => &format!(r#"rgb {}"#, from_argb(*a, *r, *g, *b)),
+			VariableRGBInteger(_) => "rgb variable",
+			VariableARGBInteger(_) => "rgb variable",
+			PaletteFracColor(v) => &format!("palette frac {v}"),
+			PaletteCBColor(v) => &format!("palette cb {v}"),
+			VariablePaletteColor(_) => "palette z",
+			SavedColorMap(s, _) => &format!("palette {s}"),
+			VariableIndex(_) => "variable",
+			Background => "background",
+			Index(n) => &format!("{}", n),
+			Black => "black",
+		};
+		String::from(str)
+	}
+
+	pub fn data(&self) -> Vec<f64>
+	{
+		match self
+		{
+			RGBString(_) => panic!("data() called on non-variable color type."),
+			RGBInteger(_, _, _) => panic!("data() called on non-variable color type."),
+			ARGBInteger(_, _, _, _) => panic!("data() called on non-variable color type."),
+			VariableRGBInteger(items) => items
+				.iter()
+				.map(|(r, g, b)| from_argb(0, *r, *g, *b) as f64)
+				.collect(),
+			VariableARGBInteger(items) => items
+				.iter()
+				.map(|(a, r, g, b)| from_argb(*a, *r, *g, *b) as f64)
+				.collect(),
+			PaletteFracColor(_) => panic!("data() called on non-variable color type."),
+			PaletteCBColor(_) => panic!("data() called on non-variable color type."),
+			VariablePaletteColor(items) => items.clone(),
+			SavedColorMap(_, items) => items.clone(),
+			Index(_) => panic!("data() called on non-variable color type."),
+			VariableIndex(items) => items.iter().map(|v| *v as f64).collect(),
+			Background => panic!("data() called on non-variable color type."),
+			Black => panic!("data() called on non-variable color type."),
+		}
+	}
+
+	pub fn is_variable(&self) -> bool
+	{
+		matches!(
+			self,
+			VariableRGBInteger(_)
+				| VariableARGBInteger(_)
+				| VariableIndex(_)
+				| VariablePaletteColor(_)
+				| SavedColorMap(_, _)
+		)
+	}
+
+	pub fn has_alpha(&self) -> bool
+	{
+		match self
+		{
+			RGBString(s) =>
+			{
+				let s = s.to_string();
+				s.starts_with("0x") && s.chars().count() == 10
```

Comment:
We know these things are ASCII, why not use `.len()` directly?

Reply:
Mostly because using `len()` on strings is discouraged. In theory the use could supply non-ascii, but then of course the whole thing wouldn't work. It probably safe, but I guess I would leave as is, but don't have a strong opinion.

Reply:
Sure, careless use of `.len()` is troublesome, but here we know precisely how many bytes the string will have.

Reply:
OK. Done